nomad/state/state_store.go

```diff
@@ -4125,6 +4133,54 @@ func (s *StateStore) upsertAllocsImpl(index uint64, allocs []*structs.Allocation
 			if alloc.Job == nil {
 				return fmt.Errorf("attempting to upsert allocation %q without a job", alloc.ID)
 			}
+
+			// Check if the alloc requires sticky volumes. If yes, find a node
+			// that has the right volume and update the task group volume
+			// claims table
+			stickyVolumes := []*structs.TaskGroupVolumeClaim{}
+			for _, tg := range alloc.Job.TaskGroups {
```

Comment:
We're iterating over the sticky volumes immediately after this, so we can create the `TaskGroupVolumeClaim` for each volume in a single pass without allocating this slice. I think that'll also let us avoid allocating the `TaskGroupVolumeClaim` if it already exists.